src/osd/PrimaryLogPG.cc

```diff
@@ -3457,24 +3458,121 @@ void PrimaryLogPG::cancel_manifest_ops(bool requeue, vector<ceph_tid_t> *tids)
+{
+  SnapSetContext* ssc = ctx->obc->ssc;
+  ceph_assert(oi.has_manifest());
+  ceph_assert(oi.soid.is_head());
+  // has snapshot
+  if (ssc && ssc->snapset.clones.size() > 0) {
+    ceph_assert(ssc);
+    dout(15) << __func__ <<  " has snapset " << dendl;
+    object_ref_delta_t refs;
+    ceph_assert(oi.manifest.is_chunked());
+    ceph_assert(!oi.manifest.is_redirect());
+
+    SnapSet& snapset = ssc->snapset;
+
+    // check adjacent clones
+    auto s = std::find(snapset.clones.begin(), snapset.clones.end(), coid.snap);
+    int index = std::distance(snapset.clones.begin(), s);
+    snapid_t l = *snapset.clones.begin(), g = CEPH_NOSNAP;
+    object_manifest_t* manifest_g = nullptr, * manifest_l = nullptr;
+    hobject_t t_oid = coid;
+
+    auto get_context = [this](const hobject_t & oid) 
+      -> ObjectContextRef {
+      ObjectContextRef obc = get_object_context(oid, false, NULL);
+      ceph_assert(obc);
+      ceph_assert(obc->obs.oi.has_manifest());
+      ceph_assert(obc->obs.oi.manifest.is_chunked());
+      return obc;
+    };
+
+    if (index) {
+      l = snapset.clones[index - 1];
+      t_oid.snap = l;
+      manifest_l = &get_context(t_oid)->obs.oi.manifest;
+    } 
+
+    if (*s != snapset.clones.back() && s != snapset.clones.end()) {
+      g = snapset.clones[index + 1];
+      t_oid.snap = g;
+      manifest_g = &get_context(t_oid)->obs.oi.manifest;
+    } else if (*s == snapset.clones.back()) {
+      manifest_g = &oi.manifest;
+    }
+    
+    if (!manifest_g) {
+      oi.manifest.calc_refs_to_drop_on_removal(manifest_g, manifest_l, refs);
+    } else {
+      get_context(coid)->obs.oi.manifest.calc_refs_to_drop_on_removal(manifest_g, manifest_l, refs);
+    }
+
+    if (!refs.is_empty()) {
+      ctx->register_on_commit(
+	[oi, ctx, this, refs](){
+	  dec_refcount(ctx->obc, oi, refs);
+      });
+    }
+    return;
+  }
+
+  // no snapshot
+  if ((oi.flags & object_info_t::FLAG_REDIRECT_HAS_REFERENCE) && oi.manifest.is_redirect()) {
+    ctx->register_on_commit(
+      [oi, ctx, this](){
+      object_locator_t target_oloc(oi.manifest.redirect_target);
+      refcount_manifest(ctx->obc, oi.soid, target_oloc, oi.manifest.redirect_target, 
+			SnapContext(), refcount_t::DECREMENT_REF, NULL);
+    });
+  } else if (oi.manifest.is_chunked()) {
+    ctx->register_on_commit(
+      [oi, ctx, this](){
+      for (auto p : oi.manifest.chunk_map) {
+	if (p.second.has_reference()) {
+	  object_locator_t target_oloc(p.second.oid);
+	  refcount_manifest(ctx->obc, oi.soid, target_oloc, p.second.oid, 
+			    SnapContext(), refcount_t::DECREMENT_REF, NULL);
+	}
+      }
+    });
+  } else {
+    ceph_abort_msg("unrecognized manifest type");
+  }
+}
+
+void PrimaryLogPG::refcount_manifest(ObjectContextRef obc, hobject_t src_soid, object_locator_t oloc,
+				     hobject_t tgt_soid, SnapContext snapc, refcount_t type, RefCountCallback* cb)
```

Comment:
obc is only used to get at soid (obc->obs.oi.soid).  I think src_soid is always obc->obs.oi.soid.   tgt_soid always implies oloc, why pass both?  Every caller passes SnapContext() for snapc -- is it ever correct to use snapshots on one of these targets?

Of the above arguments, I think you *only* need src_soid, tgt_soid, type, and cb.  Why are the others present?